cryptoprotocol/src/main/java/de/upb/crypto/incentive/cryptoprotocol/protocols/user/EarnRequest.java

```diff
@@ -0,0 +1,83 @@
+package de.upb.crypto.incentive.cryptoprotocol.protocols.user;
+
+import de.upb.crypto.craco.sig.sps.eq.SPSEQSignature;
+import de.upb.crypto.craco.sig.sps.eq.SPSEQVerificationKey;
+import de.upb.crypto.incentive.cryptoprotocol.interfaces.user.EarnInterface;
+import de.upb.crypto.incentive.cryptoprotocol.model.PublicParameters;
+import de.upb.crypto.incentive.cryptoprotocol.model.Token;
+import de.upb.crypto.incentive.cryptoprotocol.model.keys.provider.ProviderPublicKey;
+import de.upb.crypto.incentive.cryptoprotocol.model.keys.user.UserSecretKey;
+import de.upb.crypto.math.interfaces.structures.GroupElement;
+import de.upb.crypto.math.serialization.Representable;
+import de.upb.crypto.math.serialization.Representation;
+import de.upb.crypto.math.serialization.annotations.v2.ReprUtil;
+import de.upb.crypto.math.serialization.annotations.v2.Represented;
+import de.upb.crypto.math.structures.zn.Zn;
+import de.upb.crypto.math.structures.zn.Zn.ZnElement;
+
+/**
+ * Class representing an earn request created by a user of the incentive system.
+ * Provides functionality to initialize the request (i.e. preparing all data that needs to be send to the provider on making the request)
+ * as well as handling the providers response.
+ * Note that not all fields of this class are part of its serialized representation (for privacy reasons).
+ */
+public class EarnRequest implements EarnInterface, Representable {
+    @Represented
+    private PublicParameters pp;
```

Comment:
We generally refrain from serializing public parameters (and other stuff the receiving party should already know). They (1) bloat the `Representation` (you need to send lots of unncessary data over the wire) and (2) if (e.g.) the pp are part of the `Representation`, the receiver *must not* trust them anyway - who knows whether the client actually sends the correct `pp` or some maliciously chosen one (with `w = 1`, which disables traceability, or maybe with a bilinear group of bruteforceable size, or ...).

So in this case, for example, we would set it up so that an `EarnRequest`'s `Representation` doesn't contain `pp` or `pk`. Instead, to reconstruct an `EarnRequest` from Representation, you'd write a constructor `EarnRequest(Representation repr, PublicParameters pp, ProviderPublicKey pk)` that sets `this.pp` and `this.pk` (if you actually want to have those here, which I personally also wouldn't, but which would be okay) and then use `pp` and `pk` to recreate the actual values that are important here, i.e. `blindedCommitment`, `blindedCertificate` (and `k`). Note that to restore `blindedCommitment`, you need `pp`'s `Group` and to restore `blindedCertificate` you need the `SPSEQScheme` (also from `pp`, I think).

You can write the serialization/deserialization code manually (it's not actually that bad) or you can use the annotation framework (as you're currently doing). For the latter, you'll need to tell the framework where to find the `Group` for `blindedCommitment` and the `SPSEQScheme` for `blindedCertificate` (to tell it that, use `ReprUtil::register(...)`)

Generally, I recommend taking a look at https://upbcuk.github.io/docs/representations.html for more details.